crates/ty_python_semantic/src/types.rs

```diff
@@ -8993,16 +8990,136 @@ impl<'db> ConstructorCallError<'db> {
+    /// types is not described in the typing spec, but this expansion to gradual types is
+    /// sound and consistent with the principles laid out in the spec. This definition
+    /// does mean the subtyping relation is not reflexive for non-fully-static types
+    /// (e.g. `Any` is not a subtype of `Any`).
+    ///
+    /// [fully static]: https://typing.python.org/en/latest/spec/glossary.html#term-fully-static-type
+    /// [materialization]: https://typing.python.org/en/latest/spec/glossary.html#term-materialize
     Subtyping,
+
+    /// The "assignability" relation.
+    ///
+    /// The assignability relation between two types `A` and `B` dictates whether a
+    /// type checker should emit an error when a value of type `B` is assigned to a
+    /// variable declared as having type `A`.
+    ///
+    /// For a pair of [fully static] types `A` and `B`, the assignability relation
+    /// between `A` and `B` is the same as the subtyping relation.
+    ///
+    /// Between a pair of `C` and `D` where either `C` or `D` is not fully static, the
+    /// assignability relation may be more permissive than the subtyping relation. `D`
+    /// can be said to be assignable to `C` if *any* possibly fully static [materialization]
+    /// of `D` is a subtype of *any* possible fully static materialization of `C`.
+    ///
+    /// For example, `Any` is not a subtype of `int`, because there are possible
+    /// materializations of `Any` (e.g., `str`) that are not subtypes of `int`.
+    /// `Any` is *assignable* to `int`, however, as there are *some* possible materializations
+    /// of `Any` (such as `int` itself!) that *are* subtypes of `int`. `Any` cannot even
+    /// be considered a subtype of itself, as two separate uses of `Any` in the same scope
+    /// might materialize to different types between which there would exist no subtyping
+    /// relation; nor is `Any` a subtype of `int | Any`, for the same reason. Nonetheless,
+    /// `Any` is assignable to both `Any` and `int | Any`.
+    ///
+    /// While `Any` can materialize to anything, the presence of `Any` in a type does not
+    /// necessarily make it assignable to everything. For example, `list[Any]` is not
+    /// assignable to `int`, because there are no possible fully static types we could
+    /// substitute for `Any` in this type that would make it a subtype of `int`. For the
+    /// same reason, a union such as `str | Any` is not assignable to `int`.
+    ///
+    /// [fully static]: https://typing.python.org/en/latest/spec/glossary.html#term-fully-static-type
+    /// [materialization]: https://typing.python.org/en/latest/spec/glossary.html#term-materialize
     Assignability,
+
+    /// The "union simplification" relation.
+    ///
+    /// The union simplification relation dictates whether the union `A | B` can be
+    /// pragmatically simplified to the type `A` without downstream consequences on ty's
+    /// inference of types elsewhere.
+    ///
```

Comment:
I think the name is a bit unfortunate because it's hard to formulate other terms from it. I'll say that `D` is union-simplifiable with regards to `C` if the relation holds between D and C.

Then, I think the formal formulation can be that `D union-simplifiable C` if `C` is equivalent to `C | D`. Equivalence, as you probably discuss somewhere, for fully static types means that the sets of values described by both types are the same; for dynamic types it means that the top and bottom materializations of both types are equivalent.

Reply:
> I looked at the code a bit more and maybe "redundant with" is a good name. `C` is redundant with `D` if the union `C | D` can be safely simplified to `D`.

Hmm, maybe `TypeRelation::UnionRedundancy`? And we can say `C` is union-redundant with `D` if the union `C | D` can be safely simplified to `D`

Reply:
it does, yes, fair point

Reply:
Though I think the relation is reversed for intersection, right? If we use `≤R` for the new redundancy relation, and `=G` for gradual equivalence, then its

```
C ≤U D <=> (C | D) =G D
C ≤U D <=> (C & D) =G C
```

Mnemonics: the smaller thing is on the left of the relation; union makes things bigger; intersection makes things smaller

(That doesn't change Jelle's point that we don't need to include "union" in the name of the relation since it's relevant to intersections too)

Reply:
@dcreager, are you able to think of any test cases I could add for intersections? I'm struggling to come up with any, and I can't tell if this is because my head is going fuzzy with all the set theory or because our DNF representation of unions and intersections actually makes the question redundant (🥁) when it comes to intersections

Reply:
> re you able to think of any test cases I could add for intersections (that don't already pass with the logic we have on `main`)?

I can't think of any either. We can add those as a follow-on if we encounter any.

> So it does seem like it's useful (purely from a performance perspective, if nothing else) to apply it to both unions and intersections, even if (unlike with unions) this doesn't lead to any semantic improvement over what we have on `main`.

Agreed